tanjun/parsing.py

```diff
@@ -31,8 +31,321 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 from __future__ import annotations
 
-__all__: typing.Sequence[str] = []
+__all__: typing.Sequence[str] = ["argument", "option", "Parameter", "ShlexParser"]
 
+import asyncio
+import itertools
+import shlex
 import typing
 
+from hikari import undefined
+
+from tanjun import errors
 from tanjun import traits
+
+
+class ShlexTokenizer:
+    __slots__: typing.Sequence[str] = ("__arg_buffer", "__last_name", "__options_buffer", "__shlex")
+
+    def __init__(self, content: str, /) -> None:
+        self.__arg_buffer: typing.MutableSequence[str] = []
+        self.__last_name: typing.Optional[str] = None
+        self.__options_buffer: typing.MutableSequence[typing.Tuple[str, typing.Optional[str]]] = []
+        self.__shlex = shlex.shlex(content, posix=True)
+        self.__shlex.whitespace = " "
+        self.__shlex.whitespace_split = True
+
+    def collect_options(self) -> typing.Mapping[str, typing.Sequence[typing.Optional[str]]]:
+        results: typing.MutableMapping[str, typing.MutableSequence[typing.Optional[str]]] = {}
+
+        while (option_ := self.next_option()) is not None:
+            name, value = option_
+
+            if name not in results:
+                results[name] = []
+
+            results[name].append(value)
+
+        return results
+
+    def iter_arguments(self) -> typing.Iterator[str]:
+        while (argument_ := self.next_argument()) is not None:
+            yield argument_
+
+    def next_argument(self) -> typing.Optional[str]:
+        if self.__arg_buffer:
+            return self.__arg_buffer.pop(0)
+
+        while isinstance(value := self.__seek_shlex(), tuple):
+            self.__options_buffer.append(value)
+
+        return value
+
+    def next_option(self) -> typing.Optional[typing.Tuple[str, typing.Optional[str]]]:
+        if self.__options_buffer:
+            return self.__options_buffer.pop(0)
+
+        while isinstance(value := self.__seek_shlex(), str):
+            self.__arg_buffer.append(value)
+
+        return value
+
+    def __seek_shlex(self) -> typing.Union[str, typing.Tuple[str, typing.Optional[str]], None]:
+        option_name = self.__last_name
+        for value in self.__shlex:
+            is_option = value.startswith("-")
```

Comment:
Do we enforce that option names start with “-“ and argument keys dont

Reply:
Actually that’s a good point, both option and argument should have “key”, where “names” is unique to options